llvm/lib/Frontend/Offloading/PropertySet.cpp

```diff
@@ -0,0 +1,95 @@
+#include "llvm/Frontend/Offloading/PropertySet.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/MemoryBufferRef.h"
+
+namespace llvm::offloading {
+
+void writePropertiesToJSON(const PropertySetRegistry &PSRegistry,
+                           raw_ostream &Out) {
+  json::OStream J(Out);
+  J.object([&] {
+    for (const auto &[CategoryName, PropSet] : PSRegistry) {
+      J.attributeObject(CategoryName, [&] {
+        for (const auto &[PropName, PropVal] : PropSet) {
+          switch (PropVal.index()) {
+          case 0:
+            J.attribute(PropName, std::get<uint32_t>(PropVal));
+            break;
+          case 1:
+            J.attributeArray(PropName, [&] {
+              for (const auto &Byte : std::get<ByteArray>(PropVal)) {
+                J.value(Byte);
+              }
+            });
+            break;
+          default:
+            llvm_unreachable("unsupported property type");
+          }
+        }
+      });
+    }
+  });
+}
+
+template <typename... Ts> auto createStringErrorV(Ts &&...Args) {
+  return createStringError(formatv(Args...));
+}
+
+Expected<PropertySetRegistry> readPropertiesFromJSON(MemoryBufferRef Buf) {
+  PropertySetRegistry Res;
+  Expected<json::Value> V = json::parse(Buf.getBuffer());
+  if (auto E = V.takeError())
+    return E;
+
+  const json::Object *O = V->getAsObject();
+  if (!O)
+    return createStringErrorV(
+        "error while deserializing property set registry: "
+        "expected JSON object, got {0}",
+        *V);
+
+  for (const auto &[CategoryName, Value] : *O) {
+    const json::Object *PropSetVal = Value.getAsObject();
+    if (!PropSetVal)
+      return createStringErrorV("error while deserializing property set {0}: "
+                                "expected JSON array, got {1}",
+                                CategoryName.str(), Value);
+
+    PropertySet &PropSet = Res[CategoryName.str()];
+    for (const auto &[PropName, PropValueVal] : *PropSetVal) {
+      PropertyValue Prop;
+      auto PropertyParseError = [&](auto &&...MsgArgs) {
```

Comment:
This takes too much effort to read the loop body with this lambda.
If you extract the loop body in a separate function than it would not be so diffucult. Like the following:
```cpp
Expected<PropertyValue> retrievePropertyValueFromJSONObject(const json::Object *O);
```

Then you could stack errors in a simple way.